source/common/common/shared_hash_map.h

```diff
@@ -0,0 +1,373 @@
+#include "absl/strings/string_view.h"
+#include "fmt/format.h"
+
+namespace Envoy {
+
+/**
+ * Initialization parameters for SharedHashMap. The options are duplicated
+ * to the control-block after init, to aid with sanity checking when attaching
+ * an existing memory segment.
+ */
+struct SharedHashMapOptions {
+  std::string toString() const {
+    return fmt::format("capacity={}, max_key_size={}, num_slots={}", capacity, max_key_size,
+                       num_slots);
+  }
+  bool operator==(const SharedHashMapOptions& that) const {
+    return capacity == that.capacity && max_key_size == that.max_key_size &&
+           num_slots == that.num_slots;
+  }
+  bool operator!=(const SharedHashMapOptions& that) const { return !(*this == that); }
+
+  uint32_t capacity;     // how many values can be stored.
+  uint32_t max_key_size; // how many bytes of string can be stored.
+  uint32_t num_slots;    // determines speed of hash vs size efficiency.
+};
+
+/**
+ * Implements hash_map<string, Value> without using pointers, suitable
+ * for use in shared memory. Users must commit to capacity, max_key_size,
+ * and num_slots at construction time, and to the payload value at compile-time.
+ *
+ * This map may also be suitable for persisting a hash-table to long term storage,
+ * but not across machine architectures, as it doesn't use network byte order for
+ * storing ints.
+ */
+template <class Value> class SharedHashMap : public Logger::Loggable<Logger::Id::config> {
+public:
+  /**
+   * Sentinal used for next_cell links to indicate end-of-list.
+   */
+  static const uint32_t Sentinal = 0xffffffff;
+
+  /**
+   * Constructs a map control structure given a set of options, which cannot be changed.
+   * Note that the map dtaa itself is not constructed when the object is constructed.
+   * After the control-structure is constructed, the number of bytes can be computed so
+   * that a shared-memory segment can be allocated and passed to init() or attach().
+   */
+  SharedHashMap(const SharedHashMapOptions& options)
+      : options_(options), cells_(nullptr), control_(nullptr), slots_(nullptr) {}
+
+  /**
+   * Returns the numbers of byte required for the hash-table, based on
+   * the control structure. This must be used to allocate the
+   * backing-store (eg) in shared memory, which we do after
+   * constructing the object with the desired sizing.
+   */
+  size_t numBytes() const {
+    size_t size =
+        cellOffset(options_.capacity) + sizeof(Control) + options_.num_slots * sizeof(uint32_t);
+    return align(size);
+  }
+
+  /**
+   * Attempts to attach to an existing shared memory segment. Does a (relatively) quick
+   * sanity check to make sure the options copied to the provided memory match, and also
+   * that the slot, cell, and key-string structures look sane.
+   *
+   * Note that if mutex is in a locked state at the time of attachment, this function
+   * can hang.
+   */
+  bool attach(uint8_t* memory) {
```

Comment:
For extra error checking, please pass the length of memory along with the pointer, and validate that it's the expected length

Reply:
done